### DISCUSSION TITLE
PID battery.

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#611] [Q&A] [answered]
Hi.

I am writing to you, since I would like to know if in the case of the PID control you have programmed a compensator for the use of the battery?, since even when the battery is about to run out, said controller works better than the Mellinger.

I would appreciate your help.

Regards.

**Comment by hmllr** [comment, marked as answer]
Hi :)
Yes, there is battery compensation - however, it is in the stabilizer (https://github.com/bitcraze/crazyflie-firmware/blob/2023.02/src/modules/src/stabilizer.c#L218) so executed no matter which controller you use.
The values are capped to uint16 though, so I could imagine a more aggressive controller having more issues with lower battery levels, as you just cannot provide the same thrust anymore.
Hope this helps,
Hanna

**Reply by Wallat1** [reply]
Hi.

Thanks for your help. So to conclude, of the four controllers, the PID is the one that works best with low battery levels?.

Regards.

**Reply by hmllr** [reply]
We never actually did a comparison - it would be an interesting thing to know though!
We would guess that the PID controller is most stable with low battery levels, but no guarantees on that ;)

**Reply by Wallat1** [reply]
Hi.

Thank you very much for your help.

Regards.
